pilot/pkg/model/gateway.go

```diff
@@ -45,24 +56,104 @@ func MergeGateways(gateways ...Config) *MergedGateway {
+					}
+					rdsName := getRDSRouteName(s)
+					rdsRouteConfigNames[rdsName] = append(rdsRouteConfigNames[rdsName], s)
+				} else {
+					// We have duplicate port. Its not in plaintext servers. So, this has to be in TLS servers
+					// Check if this is also a HTTP server and if so, ensure uniqueness of port name
+					if isHTTPServer(s) {
+						rdsName := getRDSRouteName(s)
+						// both servers are HTTPS servers. Make sure the port names are different so that RDS can pick out individual servers
+						// WE cannot have two servers with same port name because we need the port name to distinguish one HTTPS server from another
+						// WE cannot merge two HTTPS servers even if their TLS settings have same path to the keys, because we don't know if the contents
+						// of the keys are same. So we treat them as effectively different TLS settings.
+						if _, exists := rdsRouteConfigNames[rdsName]; exists {
+							log.Debugf("skipping server on gateway %s port %s.%d.%s: non unique port name for HTTPS port",
+								spec.Name, s.Port.Name, s.Port.Number, s.Port.Protocol)
+							continue
+						}
+						rdsRouteConfigNames[rdsName] = []*networking.Server{s}
+					}
+
+					// We have another TLS server on the same port. Can differentiate servers using SNI
+					tlsServers[s.Port.Number] = append(tlsServers[s.Port.Number], s)
 				}
-				servers[s.Port.Number] = append(ss, s)
 			} else {
-				servers[s.Port.Number] = []*networking.Server{s}
+				gatewayPorts[s.Port.Number] = true
+				if isTLSServer(s) {
+					tlsServers[s.Port.Number] = []*networking.Server{s}
+				} else {
+					plaintextServers[s.Port.Number] = []*networking.Server{s}
+				}
+
+				if isHTTPServer(s) {
+					rdsRouteConfigNames[getRDSRouteName(s)] = []*networking.Server{s}
+				}
 			}
 			log.Debugf("MergeGateways: gateway %q merged server %v", name, s.Hosts)
 		}
 	}
 
+	// Concatenate both sets of servers
+	for p, v := range tlsServers {
+		servers[p] = v
+	}
+	for p, v := range plaintextServers {
+		servers[p] = v
+	}
+
 	return &MergedGateway{
-		Names:   names,
-		Servers: servers,
+		Names:               names,
+		Servers:             servers,
+		RDSRouteConfigNames: rdsRouteConfigNames,
+	}
+}
+
+func isTLSServer(server *networking.Server) bool {
+	if server.Tls != nil && !ParseProtocol(server.Port.Protocol).IsHTTP() {
+		return true
 	}
+	return false
+}
+
+func isHTTPServer(server *networking.Server) bool {
+	protocol := ParseProtocol(server.Port.Protocol)
+	if protocol.IsHTTP() {
+		return true
+	}
+
+	if protocol == ProtocolHTTPS && server.Tls != nil && server.Tls.Mode != networking.Server_TLSOptions_PASSTHROUGH {
+		return true
+	}
+
+	return false
+}
+
+func getRDSRouteName(server *networking.Server) string {
+	protocol := ParseProtocol(server.Port.Protocol)
+	if protocol.IsHTTP() {
+		return fmt.Sprintf("http.%d", server.Port.Number)
+	}
+
+	if protocol == ProtocolHTTPS && server.Tls != nil && server.Tls.Mode != networking.Server_TLSOptions_PASSTHROUGH {
```

Comment:
Why passthrough is different ? I would assume the routes on the plain text case can still be different from rules on the https.

On the other side - I'm not sure I understand why we allow https and http to have different routes.

Reply:
Added comments. Passthrough is a TCP proxy. No HTTP connection manager there. Its simply looking at SNI and forwarding to backend without decoding payload